tools/coq_dune.ml

```diff
@@ -0,0 +1,210 @@
+(* ocamlfind ocamlopt -linkpkg -package str coq_dune.ml -o coq_dune.opt *)
+open Format
+
+(* Keeping this file self-contained as it is a "bootstrap" utility *)
+(* Is OCaml missing these basic functions in the stdlib?           *)
+module Aux = struct
+
+  let option_iter f o = match o with
+    | Some x -> f x
+    | None -> ()
+
+  let option_cata d f o = match o with
+    | Some x -> f x
+    | None -> d
+
+  let list_compare f = let rec lc x y = match x, y with
+      | [], [] -> 0
+      | [], _ -> -1
+      | _, [] -> 1
+      | x::xs, y::ys -> let r = f x y in if r = 0 then lc xs ys else r
+    in lc
+
+  let rec pp_list pp sep fmt l = match l with
+    | []  -> ()
+    | [l] -> fprintf fmt "%a" pp l
+    | x::xs -> fprintf fmt "%a%a%a" pp x sep () (pp_list pp sep) xs
+
+  let rec pmap f l = match l with
+    | []  -> []
+    | x :: xs ->
+      begin match f x with
+        | None -> pmap f xs
+        | Some r -> r :: pmap f xs
+      end
+
+  let sep fmt () = fprintf fmt "@;"
+
+  module DirOrd = struct
+    type t = string list
+    let compare = list_compare String.compare
+  end
+
+  module DirMap = Map.Make(DirOrd)
+
+  (* Slower version of DirMap.update, waiting for OCaml 4.06.0 *)
+  let dirmap_update key f map =
+    match begin
+      try f (Some (DirMap.find key map))
+      with Not_found -> f None
+    end with
+    | None   -> DirMap.remove key map
+    | Some x -> DirMap.add key x map
+
+  let add_map_list key elem map =
+    (* Move to Dirmap.update once we require OCaml >= 4.06.0 *)
+    dirmap_update key (fun l -> Some (option_cata [elem] (fun ll -> elem :: ll) l)) map
+
+end
+
+open Aux
+
+type vodep = {
+  target: string;
+  deps : string list;
+}
+
+type ldep = | VO of vodep | ML4 of string | MLG of string
+type ddir = ldep list DirMap.t
+
+(* Filter `.vio` etc... *)
+let filter_no_vo =
+  List.filter (fun f -> Filename.check_suffix f ".vo")
+
+(* We could have coqdep to output dune files directly *)
+
+(* Fix once we move to OCaml >= 4.06.0 *)
+let list_init len f =
+  let rec init_aux i n f =
+    if i >= n then []
+    else let r = f i in r :: init_aux (i+1) n f
+  in init_aux 0 len f
+
+let gen_sub n =
+  (* String.concat "/" List.(init n (fun _ -> "..")) ^ "/" *)
```

Comment:
you can write
```
let gen_sub n =
  let module List = struct let rec init = ... end in (* XXX : kill on OCaml >= 4.06 *)
  String.concat "/" (List.int ....)
```

Reply:
I don't have a strong opinion on the solution I propose, but having code in comments that is supposed to be uncommented later on is a bad idea: the actual code could change, and nobody is going to fix the comment, so you may end up uncommenting old code.
What I wanted to achieve is code that stays the same.
You can also do the following if you prefer.
```
module List = struct include List let rec init .... end (* XXX remove on OCaml 4.06 *)
```

Reply:
Will this trigger a warning when compiled with more recent OCaml?

Reply:
We have plenty of code in that situation sadly but it is not an easy problem to solve, other than duplicating many modules.

Reply:
If the problem is a warning you can suppress it using an attribute.
I don't see why it is hard to solve in this case.